### PR TITLE
Adding metadata

### PR DESCRIPTION
**Pull request by carvilsi**
Removing old way to deal with metadata related with apps and providing a more general way to do this.
Related with:
CMDB-96
WE-33
WE-34

I certify that:
- [x] The automated build passes on the branch this pull request is from
- [x] My change is documented in the Changelist.md
- [x] There is a matching weaver-sdk-js pull request that uses this new server
- [x] Any new dockerimages this build depends on are properly tagged on github
- [x] I shall not merge until the travis build is green and the PR is accepted by
  another developer
- [x] I shall promptly fix/reply to comments, as I know PRs are not :fire: and
  forget
- [x] I shall keep my branch updated with any intermediate changes to the
  target branch
- [x] This is a quality PR
- [x] I got to the end of this list

### CODE REVIEW

File: src/project/ProjectService.coffee [hunk 1]
```diff
@@ -47,15 +47,29 @@ class ProjectService extends LokiService
...
+      project[bundleKey] = {}
+    project[bundleKey][key] = metadata 
     @projects.update(project)
 
-  removeApp: (user, project, appName) ->
+  removeMetadata: (user, project, bundleKey, key) ->
     project = @_getProjectToEdit(user, project)
-    delete project.apps[appName]
-    @projects.update(project)
+    if project[bundleKey]?[key]
```
Comment (bastbijl) [inline src/project/ProjectService.coffee]: there's a `?` missing at the end

File: src/project/ProjectService.coffee [hunk 2]
```diff
@@ -47,15 +47,29 @@ class ProjectService extends LokiService
...
+      throw {code: -1, message: "Not posible to delete metadata for project #{project.id} on bundle #{bundleKey} or key #{key}"}
+
+  getMetadata: (user, project, bundleKey, key) ->
+    project = @_getProjectToEdit(user, project)
+    if bundleKey? and key? and !project[bundleKey]?[key]?
+      throw {code: -1, message: "No metadata on project #{project.name} for bundleKey #{bundleKey} or key #{key}"}
+    if key? and bundleKey?
+      project[bundleKey][key]
+    else if !key? and bundleKey?
+      project[bundleKey]
```
Comment (bastbijl) [inline src/project/ProjectService.coffee]: it feels there is an `else` situation that is now not checked

File: Changelist.md [hunk 3]
```diff
@@ -1,5 +1,9 @@
 # Changelist
 
+## 3.11.1-beta.0
```
Comment (gvdent) [inline Changelist.md]: When updating a version:
- Major for backward compatibility breaking changes (try not to make these)
- Minor for new features (this is generally the default)
- Patch for bugfixes/performance improvements

This does not fit the criteria for a bugfix

File: src/project/ProjectCtrl.coffee [hunk 4]
```diff
@@ -63,16 +63,22 @@ bus.private('project.isfrozen').retrieve('project').on((req, project) ->
   ProjectService.isFrozen(project)
 )
 
-bus.private('project.app.add').retrieve('user', 'project').require('appName','appMetadata').on((req, user, project, appName, appMetadata) ->
+bus.private('project.metadata.add').retrieve('user', 'project').require('bundleKey','key','data').on((req, user, project, bundleKey, key, data) ->
```
Comment (gvdent) [inline src/project/ProjectCtrl.coffee]: This breaks backward compatibility, so older SDK versions will not be able to communicate with this weaver-server. This means that all clients should install new Weaver Desktop's and all plugins will need updating. Which implies a lot of headaches for developers across the board. 

Please make this backward compatible. So add the new routes, and implement the old routes using functionality of the new.

File: src/project/ProjectCtrl.coffee [hunk 5]
```diff
@@ -63,16 +63,22 @@ bus.private('project.isfrozen').retrieve('project').on((req, project) ->
...
+bus.private('project.metadata.remove').retrieve('user', 'project').require('bundleKey', 'key').on((req, user, project, bundleKey, key) ->
   AclService.assertProjectFunctionPermission(user, project, 'project-administration')
-  logger.code.info "User #{ul(req, user)} removes app #{appName} from project id: #{project.id}"
-  ProjectService.removeApp(user, project, appName)
+  logger.code.info "User #{ul(req, user)} removes metadata #{key} from project id: #{project.id} at bundleKey: #{bundleKey}"
+  ProjectService.removeMetadata(user, project, bundleKey , key)
+)
+
+bus.private('project.metadata.get').retrieve('user', 'project').require('bundleKey').optional('key').on((req, user, project, bundleKey, key) ->
+  AclService.assertProjectFunctionPermission(user, project, 'project-administration')
```
Comment (gvdent) [inline src/project/ProjectCtrl.coffee]: Are you sure that this needs 'project-administration' permission? This means that regular users will not be able to ever retrieve meta data, so a non-admin user can never get a list of possible apps etc.

File: src/routes.coffee [hunk 6]
```diff
@@ -68,8 +68,9 @@ route.private.GET  "project"                 # Get a list of projects
 route.private.GET  "project.freeze"          # Freezes a project
 route.private.GET  "project.unfreeze"        # Unfreezes project
 route.private.GET  "project.isfrozen"        # Checks the status frozen/nonFrozen for a project
-route.private.POST "project.app.add"         # Add compatible app to project
-route.private.POST "project.app.remove"      # Remove compatible app to project
```
Comment (gvdent) [inline src/routes.coffee]: These two lines shouldn't be removed.

File: src/project/ProjectService.coffee (hunk 1 above)
Comment (carvilsi) [reply src/project/ProjectService.coffee]: Yes 👍

File: src/project/ProjectService.coffee (hunk 2 above)
Comment (gvdent) [reply src/project/ProjectService.coffee]: Indeed, this is not clear at all and there is some cases being missed, (e.g. having a bundleKey but it not being defined results in an `undefined` result while having both a bundleKey and key will result in an exception if no results are found).

Please rewrite to be more clear and obvious correct, in general prevent duplicate checks in if statements (here bundleKey and key are checked 3 times for existence), nest instead.

File: src/project/ProjectCtrl.coffee (hunk 4 above)
Comment (carvilsi) [reply src/project/ProjectCtrl.coffee]: The backwards compatibility is already take in account, only weaver-electron was using this and even there was never used on production. In the other hand this is already update in another PR for weaver-electron.

File: src/project/ProjectCtrl.coffee (hunk 5 above)
Comment (carvilsi) [reply src/project/ProjectCtrl.coffee]: hmmmm need to check it! 👍

File: src/routes.coffee (hunk 6 above)
Comment (carvilsi) [reply src/routes.coffee]: The backward compatibility is not a problem in this case, the lines can be removed.

File: src/project/ProjectCtrl.coffee (hunk 5 above)
Comment (carvilsi) [reply src/project/ProjectCtrl.coffee]: I think that this 'project-administration' permission check to read is not need here! 👍

**Comment by gvdent** [comment]
Backward compatibility is not something that can be 'taken into account' by claiming there is no software that is using it on a publicly published service or api. That is analogous to claiming there is no teapot orbiting the sun between earth and mars.

We do not know all software using our server or SDK, that is impossible. Publishing backward compatibility changes under anything other than major releases could break someone's software and even the possibility is a cardinal sin imo.

This is published publicly, we as engineers should take the responsibility to create an API that is as solid and stable as possible. Changing the API is a last resort and should be done as little as possible. In this case the effort required to maintain backward compatibility in my opinion outweighs both releasing a breaking change under a less-than-major version, and upgrading the major version which means all software in the world needs to be updated.

**Pull request merged**